File: application/src/main/java/com/smtm/application/security/infrastructure/SmtmUserDetailsService.java
package com.smtm.application.security.infrastructure;

import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;

// In order to make Spring Security magic work properly, it seems that this should be the only implementation of UserServiceDetails interface
// It cannot be exposed as @Bean method with parameters or I can't do it.
@Service
public class SmtmUserDetailsService implements UserDetailsService {

    private final DbUsersRepository dbUsersRepository;

    public SmtmUserDetailsService(DbUsersRepository dbUsersRepository) {
        this.dbUsersRepository = dbUsersRepository;
    }

    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        return dbUsersRepository.findByEmail(username);
    }
}

File: application/src/main/kotlin/com/smtm/application/security/SmtmUserDetailsService.kt
package com.smtm.application.security

import com.smtm.infrastructure.persistence.users.DbUsersRepository
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.stereotype.Service

// In order to make Spring Security magic work properly, it seems that this should be the only implementation of UserServiceDetails interface
// It cannot be exposed as @Bean method with parameters or I can't do it.
@Service
class SmtmUserDetailsService(private val dbUsersRepository: DbUsersRepository) : UserDetailsService {

    override fun loadUserByUsername(username: String): UserDetails? = dbUsersRepository.findByEmail(username)
}
